### DISCUSSION TITLE
Simpler style for declaring shared variables

### DISCUSSION THREAD

**Discussion by listrophy** [Quick/Quick#1078] [Ideas]
Apple's builtin XCTest framework has a much simpler way to instantiate shared variables. This mechanism does not work with Quick.

Given the simple class:

``` swift
class Mutator {
    var mutated: Bool
    init() {
        mutated = false
    }

    func mutate() {
        mutated = true
    }
}
```

I'd like to see a way to do this:

``` swift
// Please note that this style does not work, but I'd like it to!

class MyQuickTest: QuickSpec {
    override func spec() {

        describe("Mutator") {
            let subject = Mutator()

            it("test 1") {
                expect(subject.mutated).to(beFalse())
                subject.mutate()
                expect(subject.mutated).to(beTrue())
            }

            it("test 2") {
                expect(subject.mutated).to(beFalse()) // currently fails because `subject` is same mutated object from "test 1"
                subject.mutate()
                expect(subject.mutated).to(beTrue())
            }
        }
    }
}
```

Apple does this in XCTest by instantiating a new object for each test method.

Currently, I'd have to do this, which seems overly verbose and has those gross `!` optional-smashers everywhere:

``` swift
// undesirables:
//   1. must declare shared variable as "var" and not "let".
//   2. must use `beforeEach` to instantiate before each test run
//   3. must use "!" (or worse `if let`) to access underlying value since it's now an optional

class MyQuickTest: QuickSpec {
    override func spec() {

        describe("Mutator") {
            var subject: Mutator?
            beforeEach {
                subject = Mutator()
            }

            it("test 1") {
                expect(subject!.mutated).to(beFalse())
                subject!.mutate()
                expect(subject!.mutated).to(beTrue())
            }

            it("test 2") {
                expect(subject!.mutated).to(beFalse())
                subject!.mutate()
                expect(subject!.mutated).to(beTrue())
            }
        }
    }
}
```

**Comment by modocache** [comment]
I definitely agree that your "undesirables" are not desirable. This issue deserves further consideration--although this syntax wasn't very painful in Objective-C, it certainly is in Swift. Thanks for prompting me to think about this problem, @listrophy! :+1: 

Obviously this isn't a great solution, but for what it's worth, I always use implicitly unwrapped types, i.e.: `Mutator!`. :sweat: 

Question: does the following work?

``` swift
class MyQuickTest: QuickSpec {
  let subject = Mutator()

  override func spec() {
    describe("Mutator") {
      it("test 1") {
        expect(subject.mutated).to(beFalse())
        subject.mutate()
        expect(subject.mutated).to(beTrue())
      }

      it("test 2") {
        expect(subject.mutated).to(beFalse())
        subject.mutate()
        expect(subject.mutated).to(beTrue())
      }
    }
  }
}
```

**Comment by paulyoung** [comment]
I imagine that one of these will fail since when the second runs `subject.mutated` will already be `true`.

**Comment by modocache** [comment]
Will it, though? As @listrophy points out, a fresh instsance of `XCTestCase` is used every time `-[XCTestCase invokeTest]` is called. Since `QuickSpec` is a humble `XCTestCase` subclass, it's possible this might work. :bulb: 

Gonna have to actually try it locally, though! :sweat_smile:

**Comment by paulyoung** [comment]
I had to write it as follows:

``` swift
class MyQuickTest: QuickSpec {
    let subject = Mutator()

    override func spec() {
        describe("Mutator") {
            it("test 1") {
                expect(self.subject.mutated).to(beFalse())
                self.subject.mutate()
                expect(self.subject.mutated).to(beTrue())
            }

            it("test 2") {
                expect(self.subject.mutated).to(beFalse())
                self.subject.mutate()
                expect(self.subject.mutated).to(beTrue())
            }
        }
    }
}
```

The first expectation in "test 2" failed.

**Comment by listrophy** [comment]
@paulyoung that's probably because the underlying XCTest framework only sees one test method: `spec`.

Here's what I think happens:
1. XCTest suite loads up all descendants of `XCTestCase`
2. It reflects on each of those classes and determines how many test methods there are in each one. In a non-Quick environment, this is probably just "count the number of methods that start with 'test'"
3. The suite then instantiates that many instances for each class.
4. The suite then runs the tests, using a unique instance for each test.

Because the `QuickSpec` class only has `spec`, the suite only instantiates one instance of the class.

See also the [blog post](http://bendyworks.com/unit-testing-in-swift/) I wrote back in June, specifically the part under "How XCTest Handles Multiple Tests in One TestCase"

**Comment by modocache** [comment]
I may be wrong, but I think XCTest is picking up the correct number of tests here--Quick overrides `+[XCTestCase testInvocations]` to return an array of invocation objects. XCTest should correctly determine that the number of tests is `[[XCTestCase testInvocations] count]` --after all, it correctly reports the number of tests that were run, etc.

So this definitely deserves more attention, but I'm not sure that specifically is the issue here.

**Comment by briancroom** [comment]
I'm thinking about this issue now after having worked on a related issue for swift-corelibs-xctest. The reason why the example given by @paulyoung doesn't work as expected is because, although XCTest is making a new instance for each test, those instances' properties aren't referenced by the test closures at all. Instead, the test closures which are being executed have captured the property values of the test case instance which Quick itself instantiates during its setup process. (See [here](https://github.com/Quick/Quick/blob/master/Sources/Quick/QuickSpec.m#L32))

This isn't proposing a solution (yet, anyway), but I wanted to dump that here before I forget the context again!
